ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/bytes/VarInt.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+@file:Suppress("NOTHING_TO_INLINE")
+
+package io.ktor.network.quic.bytes
+
+import io.ktor.network.quic.util.*
+import io.ktor.utils.io.core.*
+
+/**
+ * Reads variable-length non-negative integer value.
+ * Executes [elseBlock] if EOF.
+ *
+ * [RFC Reference](https://www.rfc-editor.org/rfc/rfc9000.html#name-variable-length-integer-enc)
+ *
+ * @see [ByteReadPacket.readVarInt]
+ */
+internal inline fun ByteReadPacket.readVarIntOrElse(elseBlock: () -> Long): Long {
+    val value = readVarInt()
+    if (value == -1L) {
+        return elseBlock()
+    }
+    return value
+}
+
+/**
+ * Reads variable-length non-negative integer value.
+ * Returns -1 if EOF.
+ *
+ * [RFC Reference](https://www.rfc-editor.org/rfc/rfc9000.html#name-variable-length-integer-enc)
+ */
+@OptIn(ExperimentalUnsignedTypes::class)
+private fun ByteReadPacket.readVarInt(): Long {
+    val value = tryPeek().let { if (it == -1) return -1 else it }
```

Comment:
minor: I would unroll this one:
```kotlin
val value = tryPeek()
if (value == -1) return -1
```

Reply:
done. Thanks for your comments and suggestions!

Reply:
thanks for working on it!
Im willing to try working implementation in combination with rsocket-kotlin when it will be ready (even in some snapshot version).
If everything goes well, it will also allow me to do cross test it with netty-quic (via quiche) implementation!